forward/rewriting.lua

```diff
@@ -134,7 +135,64 @@ function Rewriting:process_packet(i, o)
    else
       l4_type = ip_header:next_header()
    end
-   if not (l4_type == L4_TCP or l4_type == L4_UDP) then
+   local ip_total_length = ip_header:total_length()
+   -- Check for IP fragmentation
+   local frag_off = ip_header:frag_off()
+   local mf = band(ip_header:flags(), IP_MF_FLAG) ~= 0
+   if frag_off ~= 0 or mf then
+      local src_port, dst_port = 0, 0
+      -- TODO: Redirect to spike backend pool.
+      P.free(p)
+      return
+   end
+   if l4_type == L4_GRE then
+      -- Start of IP fragment extraction and reassembly
+      -- Might want to extract this block into a separate function
+      local gre_class = ip_header:upper_layer()
+      local gre_header = datagram:parse_match(gre_class)
+      -- Checking the recursion control bit seems unnecessary here since
+      -- there is no other reason we would encounter a GRE header...?
+      if gre_header:protocol() ~= L3_IPV4 then
+         P.free(p)
+         return
+      end
+      -- Snabb thinks that IPv4 cannot be an upper layer for GRE;
+      -- so just supply the upper layer class directly.
+      -- Should we PR to Snabb?
```

Comment:
What do you mean by "Snabb thinks that IPv4 cannot be an upper layer for GRE" exactly?  This does sound like a bug in snabb.

Reply:
When we call upper_layer(), Snabb finds the protocol class for the upper layer by 1) extracting the protocol code from the header's upper-layer protocol field and 2) looking that up in a code-to-protocol-class dictionary called _ulp. The _ulp for GRE in Snabb (https://github.com/snabbco/snabb/blob/707db7f02921431b770ce7307234476d667eb93d/src/lib/protocol/gre.lua#L30) contains only Ethernet (i.e. no IPv4), so upper_layer() will not give us the IPV4 class even when the GRE header's upper-layer protocol field says it's IPv4.

Reply:
Got it.  I recommend you fix this bug in a PR to snabb; seems like it should at least contain entries for IPv4 and IPv6.

Reply:
PR here: https://github.com/snabbco/snabb/pull/1240